samcli/lib/build/workflow_config.py

```diff
@@ -2,60 +2,158 @@
     """
 
-    workflow_config_by_runtime = {
-        "python2.7": PYTHON_PIP_CONFIG,
-        "python3.6": PYTHON_PIP_CONFIG,
-        "python3.7": PYTHON_PIP_CONFIG,
-        "nodejs4.3": NODEJS_NPM_CONFIG,
-        "nodejs6.10": NODEJS_NPM_CONFIG,
-        "nodejs8.10": NODEJS_NPM_CONFIG,
-        "ruby2.5": RUBY_BUNDLER_CONFIG
+    selectors_by_runtime = {
+        "python2.7": BasicWorkflowSelector(PYTHON_PIP_CONFIG),
+        "python3.6": BasicWorkflowSelector(PYTHON_PIP_CONFIG),
+        "python3.7": BasicWorkflowSelector(PYTHON_PIP_CONFIG),
+        "nodejs4.3": BasicWorkflowSelector(NODEJS_NPM_CONFIG),
+        "nodejs6.10": BasicWorkflowSelector(NODEJS_NPM_CONFIG),
+        "nodejs8.10": BasicWorkflowSelector(NODEJS_NPM_CONFIG),
+        "ruby2.5": BasicWorkflowSelector(RUBY_BUNDLER_CONFIG),
+
+        # When Maven builder exists, add to this list so we can automatically choose a builder based on the supported
+        # manifest
+        "java8": ManifestWorkflowSelector([
+            # Gradle builder needs custom executable paths to find `gradlew` binary
+            JAVA_GRADLE_CONFIG._replace(executable_search_paths=[code_dir, project_dir])
+        ]),
     }
 
-    try:
-        return workflow_config_by_runtime[runtime]
-    except KeyError:
+    if runtime not in selectors_by_runtime:
         raise UnsupportedRuntimeException("'{}' runtime is not supported".format(runtime))
+
+    selector = selectors_by_runtime[runtime]
+
+    try:
+        config = selector.get_config(code_dir, project_dir)
+        return config
+    except ValueError as ex:
+        raise UnsupportedRuntimeException("Unable to find a supported build workflow for runtime '{}'. Reason: {}"
+                                          .format(runtime, str(ex)))
+
+
+class BasicWorkflowSelector(object):
+    """
+    Basic workflow selector that returns the first available configuration in the given list of configurations
+    """
+
+    def __init__(self, configs):
+
+        if not isinstance(configs, list):
+            configs = [configs]
+
+        self.configs = configs
+
+    def get_config(self, code_dir, project_dir):
+        """
+        Returns the first available configuration
+        """
+        return self.configs[0]
+
+
+class ManifestWorkflowSelector(BasicWorkflowSelector):
+    """
+    Selects a workflow by examining the directories for presence of a supported manifest
+    """
+
+    def get_config(self, code_dir, project_dir):
+        """
+        Finds a configuration by looking for a manifest in the given directories.
+
+        Returns
+        -------
+        samcli.lib.build.workflow_config.CONFIG
+            A supported configuration if one is found
+
+        Raises
+        ------
+        ValueError
+            If none of the supported manifests files are found
+        """
+
+        # Search for manifest first in code directory and then in the project directory.
```

Comment:
How this works and is resolved needs to be clearly documented (somewhere)

Reply:
I will add to the help text for now